render-app/src/main/java/org/janelia/alignment/spec/ResolvedTileSpecsWithMatchPairs.java

```diff
@@ -30,37 +35,101 @@ public ResolvedTileSpecsWithMatchPairs(final ResolvedTileSpecCollection resolved
+     *                        (or null to accept all pairs).
+     *
+     * @throws IllegalArgumentException
+     *   if maxZDistance < 0
+     */
+    public void normalize(final Set<String> tileIdsToKeep,
+                          final Integer maxZDistance)
             throws IllegalArgumentException {
-        if (! isNormalized) {
-            resolvedTileSpecs.resolveTileSpecs();
-            final List<CanvasMatches> normalizedMatchPairs = new ArrayList<>(matchPairs.size());
-            for (final CanvasMatches pair : matchPairs) {
-                if (resolvedTileSpecs.hasTileSpec(pair.getpId()) && (resolvedTileSpecs.hasTileSpec(pair.getqId()))) {
+
+        final Integer numberOfTilesToKeep = tileIdsToKeep == null ? null : tileIdsToKeep.size();
+
+        LOG.info("normalize: entry, process {} tiles and {} pairs with {} tileIdsToKeep and maxZDistance {}",
+                 resolvedTileSpecs.getTileCount(), matchPairs.size(), numberOfTilesToKeep, maxZDistance);
+
+        if ((maxZDistance != null) && (maxZDistance <= 0)) {
+            throw new IllegalArgumentException("maxZDistance must be >= 0 or null");
+        }
+
+        // remove unwanted tiles (which will later cause associated math pairs to be removed)
+        if (tileIdsToKeep != null) {
+
+            // track and log removal info if debugging
+            final boolean isDebugEnabled = LOG.isDebugEnabled();
+            final Set<String> beforeTileIds = isDebugEnabled ? new HashSet<>(resolvedTileSpecs.getTileIds()) : null;
+
+            resolvedTileSpecs.removeDifferentTileSpecs(tileIdsToKeep);
+
+            if (isDebugEnabled) {
+                final Set<String> afterTileIds = resolvedTileSpecs.getTileIds();
+                final int removalCount = beforeTileIds.size() - afterTileIds.size();
+                if (removalCount > 0) {
+                    beforeTileIds.removeAll(afterTileIds);
+                    LOG.debug("normalize: removed {} tiles including {}",
+                              removalCount, beforeTileIds.iterator().next());
+                }
+            }
+        }
+
+        resolvedTileSpecs.resolveTileSpecs();
+
+        // remove match pairs that are too far away in z or do not have both tile specs in this collection
+        // TODO: keep track of or return removed match pairs in case solver needs to pull adjacent tiles later
+
+        final List<CanvasMatches> normalizedMatchPairs = new ArrayList<>(matchPairs.size());
+
+        for (final CanvasMatches pair : matchPairs) {
+            final TileSpec pTileSpec = resolvedTileSpecs.getTileSpec(pair.getpId());
+            final TileSpec qTileSpec = resolvedTileSpecs.getTileSpec(pair.getqId());
+            if ((pTileSpec != null) && (qTileSpec != null)) {
+                if (maxZDistance == null) {
                     normalizedMatchPairs.add(pair);
+                } else {
+                    final int zDistance = (int) Math.abs(pTileSpec.getZ() - qTileSpec.getZ());
```

Comment:
Could this use the new `getIntegralZ()` method?

Reply:
yes - actually decided to take it one step further and add a zDistanceFrom method to TileSpec instead